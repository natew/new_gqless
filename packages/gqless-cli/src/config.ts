import { cosmiconfig } from 'cosmiconfig';
import { promises } from 'fs';
import { resolve } from 'path';
import { Boolean, Partial, Static, String, Unknown } from 'runtypes';

import type { GenerateOptions } from './generate';
import type { IntrospectionOptions } from './introspection';

type GqlessCombinedOptions = GenerateOptions & {
  /**
   * Introspection options
   */
  introspection?: IntrospectionOptions;
  /**
   * Client generation destination
   */
  destination?: string;
};

export type GqlessConfig = Required<
  Static<typeof gqlessCLIConfigRecord>
> extends Required<GqlessCombinedOptions>
  ? GqlessCombinedOptions
  : never;

const StringRecord = Unknown.withConstraint<Record<string, string>>(
  (scalars) => {
    return (
      typeof scalars === 'object' &&
      scalars != null &&
      !Array.isArray(scalars) &&
      Object.values(scalars).every((v) => typeof v === 'string')
    );
  }
);

const gqlessCLIConfigRecord = Partial({
  endpoint: String,
  enumsAsStrings: Boolean,
  scalars: StringRecord,
  react: Boolean,
  preImport: String,
  introspection: Partial({
    endpoint: String,
    headers: StringRecord,
  }),
  destination: String,
  subscriptions: Boolean,
});

export const defaultConfig = {
  endpoint: '/api/graphql',
  enumsAsStrings: false,
  react: false,
  scalars: {
    DateTime: 'string',
  },
  preImport: '',
  introspection: {
    endpoint: 'SPECIFY_ENDPOINT_OR_SCHEMA_FILE_PATH_HERE',
    headers: {},
  } as IntrospectionOptions,
  destination: './src/generated/graphql.ts',
  subscriptions: false,
};

const defaultFilePath = resolve(process.cwd(), 'gqless.config.cjs');

type DeepReadonly<T> = T extends (infer R)[]
  ? DeepReadonlyArray<R>
  : T extends Function
  ? T
  : T extends object
  ? DeepReadonlyObject<T>
  : T;

interface DeepReadonlyArray<T> extends ReadonlyArray<DeepReadonly<T>> {}

type DeepReadonlyObject<T> = {
  readonly [P in keyof T]: DeepReadonly<T[P]>;
};

export const gqlessConfigPromise: Promise<{
  filepath: string;
  config: DeepReadonly<GqlessConfig>;
}> = cosmiconfig('gqless', {
  searchPlaces: ['gqless.config.cjs', 'gqless.config.js', 'package.json'],
})
  .search()
  .then(async (config) => {
    if (!config || config.isEmpty) {
      const filepath = config?.filepath || defaultFilePath;

      const NODE_ENV = process.env['NODE_ENV'];

      if (NODE_ENV !== 'test' && NODE_ENV !== 'production') {
        const { format } = (await import('./prettier')).formatPrettier({
          parser: 'typescript',
        });
        await promises.writeFile(
          defaultFilePath,
          await format(`
          /**
           * @type {import("@dish/gqless-cli").GqlessConfig}
           */
          const config = ${JSON.stringify(defaultConfig)};
          
          module.exports = config;`)
        );
      }
      return {
        filepath,
        config: defaultConfig,
      };
    }
    return {
      config: gqlessCLIConfigRecord.check(config.config),
      filepath: config.filepath,
    };
  })
  .catch(() => ({ config: defaultConfig, filepath: defaultFilePath }));
