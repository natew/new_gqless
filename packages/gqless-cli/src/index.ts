export * from "./generate";
export * from "./introspection";
export * from "./writeGenerate";
