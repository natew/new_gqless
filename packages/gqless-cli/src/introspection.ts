import { fetch } from 'cross-fetch';
import { print } from 'graphql';

import { AsyncExecutor } from '@graphql-tools/delegate';
import { introspectSchema, wrapSchema } from '@graphql-tools/wrap';

export const getRemoteSchema = async (
  /**
   * Endpoint of the remote GraphQL API, if not specified, it points to http://localhost:3000/graphql
   */
  endpoint: string,
  /**
   * Specify configuration like headers for the introspection
   */
  { headers = {} }: { headers?: Record<string, string> } = {}
) => {
  const executor: AsyncExecutor = async ({ document, variables }) => {
    const query = print(document);
    const fetchResult = await fetch(endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      body: JSON.stringify({ query, variables }),
    });
    return fetchResult.json();
  };

  const schema = wrapSchema({
    schema: await introspectSchema(executor, {
      endpoint,
    }),
    executor,
  });

  return schema;
};
