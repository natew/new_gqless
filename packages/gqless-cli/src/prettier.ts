import {
  Options as PrettierOptions,
  resolveConfig,
  format as prettierFormat,
} from 'prettier';

const commonConfig = resolveConfig(process.cwd());

export function formatPrettier(
  defaultOptions: Omit<PrettierOptions, 'parser'> &
    Required<Pick<PrettierOptions, 'parser'>>
) {
  const configPromise = commonConfig.then((config) =>
    Object.assign({}, config, defaultOptions)
  );

  return {
    async format(input: string) {
      return prettierFormat(input, await configPromise);
    },
  };
}
