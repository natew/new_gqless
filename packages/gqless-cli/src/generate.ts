import {
  GraphQLEnumType,
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLSchema,
  parse,
  printSchema,
} from 'graphql';
import fromPairs from 'lodash/fromPairs';
import { format, Options as PrettierOptions, resolveConfig } from 'prettier';

import { parseSchemaType, ScalarsEnumsHash, Schema, Type } from '@dish/gqless';
import { codegen } from '@graphql-codegen/core';
import * as typescriptPlugin from '@graphql-codegen/typescript';

export type GenerateOptions = {
  /**
   * Overwrite the default 'queryFetcher'
   * 
   * @default
   * const queryFetcher: QueryFetcher = async function (query, variables) {
        const response = await fetch('/graphql', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            query,
            variables,
          }),
          mode: 'cors',
        });

      if (!response.ok) {
        throw new Error(`Network error, received status code ${response.status}`);
      }

      const json = await response.json();

      return json;
    };
   */
  queryFetcher?: string;
  /**
   * Add a custom string at the beginning of the file, for example, add imports.
   */
  preImport?: string;
  /**
   * Specify the types of custom scalars
   */
  scalars?: Record<string, string>;
};

export async function generate(
  schema: GraphQLSchema,
  { queryFetcher, preImport = '', scalars }: GenerateOptions = {}
) {
  const prettierConfig = resolveConfig(process.cwd());
  const codegenResult = codegen({
    schema: parse(printSchema(schema)),
    config: {},
    documents: [],
    filename: 'gqless.generated.ts',
    pluginMap: {
      typescript: typescriptPlugin,
    },
    plugins: [
      {
        typescript: {
          onlyOperationTypes: true,
          declarationKind: 'interface',
          addUnderscoreToArgsType: true,
          scalars,
          namingConvention: 'keep',
        },
      },
    ],
  });

  const config = schema.toConfig();

  const scalarsEnumsHash: ScalarsEnumsHash = {};

  const enumsNames: string[] = [];

  const inputTypeNames = new Set<string>();

  const generatedSchema: Schema = {
    query: {},
    mutation: {},
    subscription: {},
  };

  const queryType = config.query;
  const mutationType = config.mutation;
  const subscriptionType = config.subscription;

  const parseEnumType = (type: GraphQLEnumType) => {
    scalarsEnumsHash[type.name] = true;
    enumsNames.push(type.name);
  };
  const parseScalarType = (type: GraphQLScalarType) => {
    scalarsEnumsHash[type.name] = true;
  };
  const parseObjectType = (type: GraphQLObjectType, typeName = type.name) => {
    const fields = type.getFields();

    const schemaType: Record<string, Type> = {};

    Object.entries(fields).forEach(([key, value]) => {
      schemaType[key] = {
        __type: value.type.toString(),
      };

      if (value.args.length) {
        schemaType[key].__args = fromPairs(
          value.args.map((arg) => {
            return [arg.name, arg.type.toString()];
          })
        );
      }
    });

    generatedSchema[typeName] = schemaType;
  };

  const parseInputType = (type: GraphQLInputObjectType) => {
    inputTypeNames.add(type.name);
    const fields = type.getFields();

    const schemaType: Record<string, Type> = {};

    Object.entries(fields).forEach(([key, value]) => {
      schemaType[key] = {
        __type: value.type.toString(),
      };
    });

    generatedSchema[type.name] = schemaType;
  };

  config.types.forEach((type) => {
    if (
      type.name.startsWith('__') ||
      type === queryType ||
      type === mutationType ||
      type === subscriptionType
    )
      return;

    if (type instanceof GraphQLEnumType) {
      parseEnumType(type);
    } else if (type instanceof GraphQLScalarType) {
      parseScalarType(type);
    } else if (type instanceof GraphQLObjectType) {
      parseObjectType(type);
    } else if (type instanceof GraphQLInputObjectType) {
      parseInputType(type);
    }
  });

  /* istanbul ignore else */
  if (queryType) {
    parseObjectType(queryType, 'query');
  }

  if (mutationType) {
    parseObjectType(mutationType, 'mutation');
  }

  if (subscriptionType) {
    parseObjectType(subscriptionType, 'subscription');
  }

  function parseFinalType({
    pureType,
    isArray,
    nullableItems,
    isNullable,
  }: ReturnType<typeof parseSchemaType>) {
    let typeToReturn: string[] = [
      scalarsEnumsHash[pureType] ? `ScalarsEnums["${pureType}"]` : pureType,
    ];

    if (isArray) {
      typeToReturn = [
        'Array<',
        ...(nullableItems ? ['Maybe<', ...typeToReturn, '>'] : typeToReturn),
        '>',
      ];
    }

    if (isNullable) {
      typeToReturn = ['Maybe<', ...typeToReturn, '>'];
    }

    return typeToReturn.join('');
  }

  let typescriptTypes = Object.entries(generatedSchema).reduce(
    (acum, [typeKey, typeValue]) => {
      const typeName = (() => {
        switch (typeKey) {
          case 'query': {
            return 'Query';
          }
          case 'mutation': {
            return 'Mutation';
          }
          case 'subscription': {
            return 'Subscription';
          }
          default: {
            return typeKey;
          }
        }
      })();

      if (inputTypeNames.has(typeName)) return acum;

      acum += `

      export interface ${typeName} {
        ${Object.entries(typeValue).reduce((acum, [fieldKey, fieldValue]) => {
          const fieldValueProps = parseSchemaType(fieldValue.__type);
          const typeToReturn = parseFinalType(fieldValueProps);
          if (fieldValue.__args) {
            const argsEntries = Object.entries(fieldValue.__args);
            let onlyNullableArgs = true;
            const argTypes = argsEntries.reduce(
              (acum, [argKey, argValue], index) => {
                const argValueProps = parseSchemaType(argValue);
                const connector = argValueProps.isNullable ? '?:' : ':';

                if (!argValueProps.isNullable) {
                  onlyNullableArgs = false;
                }

                const argTypeValue = parseFinalType(argValueProps);

                acum += `${argKey}${connector} ${argTypeValue}`;
                if (index < argsEntries.length - 1) {
                  acum += '; ';
                }
                return acum;
              },
              ''
            );
            const connector = onlyNullableArgs ? '?:' : ':';
            acum += `
            ${fieldKey}: (args${connector} {${argTypes}}) => ${typeToReturn}`;
          } else {
            acum += `
            ${fieldKey}: ${typeToReturn}`;
          }

          return acum;
        }, '')}
      }
      `;

      return acum;
    },
    ''
  );

  typescriptTypes += `
    export interface GeneratedSchema {
      query: Query
      mutation: Mutation
      subscription: Subscription
    }
    `;

  typescriptTypes += `
    export interface ScalarsEnums extends Scalars {
      ${enumsNames.reduce((acum, enumName) => {
        acum += `${enumName}: ${enumName};`;
        return acum;
      }, '')}
    }
    `;

  queryFetcher =
    queryFetcher ||
    `
    const queryFetcher : QueryFetcher = async function (query, variables) {
        const response = await fetch("/graphql", {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            query,
            variables,
          }),
          mode: "cors",
        });
      
        if (!response.ok) {
          throw new Error(\`Network error, received status code \${response.status}\`);
        }
      
        const json = await response.json();
      
        return json;
      };
    `;

  const code = format(
    `
  ${preImport}
  import { createClient, QueryFetcher, ScalarsEnumsHash } from "@dish/gqless";

  ${await codegenResult}

  const scalarsEnumsHash: ScalarsEnumsHash = ${JSON.stringify(
    scalarsEnumsHash
  )};
  export const generatedSchema = ${JSON.stringify(generatedSchema)} as const;

  ${typescriptTypes}

  ${queryFetcher}

  export const client = createClient<GeneratedSchema>(generatedSchema, scalarsEnumsHash, queryFetcher)

  export const { query, mutation, subscription, resolved, selectFields } = client;
  `,
    Object.assign({}, await prettierConfig, {
      parser: 'typescript',
    } as PrettierOptions)
  );

  return {
    code,
    generatedSchema,
    scalarsEnumsHash,
  };
}
