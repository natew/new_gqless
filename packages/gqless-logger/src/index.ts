import parserJSON from 'prettier/parser-babel';
import parserGraphQL from 'prettier/parser-graphql';
import prettier from 'prettier/standalone';

import { createClient } from '@dish/gqless';
import { FetchEventData } from '@dish/gqless/dist/Events';

function parseGraphQL(query: string) {
  return prettier.format(query, {
    parser: 'graphql',
    plugins: [parserGraphQL],
  });
}

function cleanObject(obj: object): object {
  return JSON.parse(JSON.stringify(obj));
}

function isTruthy<V>(v: V): v is NonNullable<V> {
  return Boolean(v);
}

const format = (...parts: Array<any[] | undefined>) => {
  const texts: string[] = [];
  const styles: string[] = [];
  for (const [text, style] of parts.filter(isTruthy)) {
    texts.push(text);
    styles.push(`font-weight: normal; ${style}`);
  }

  return [`%c${texts.join('%c')}`, ...styles];
};

export interface LoggerOptions {
  /**
   * Show selections in logs
   *
   * @default
   * true
   */
  showSelections?: boolean;
  /**
   * Show cache snapshots in logs
   *
   * @default
   * true
   */
  showCache?: boolean;
  /**
   * Stringify JSON
   *
   * @default
   * false
   */
  stringifyJSON?: boolean;
}

export function createLogger(
  client: ReturnType<typeof createClient>,
  options: LoggerOptions = {}
) {
  options = Object.assign({}, options);

  options.showCache ??= true;
  options.showSelections ??= true;
  options.stringifyJSON ??= false;

  const stringifyJSONIfEnabled = <T>(v: T) => {
    if (options.stringifyJSON) {
      return prettier.format(JSON.stringify(v), {
        parser: 'json',
        plugins: [parserJSON],
      });
    }
    return v;
  };

  const eventHandler = client.eventHandler;

  async function onFetch(dataPromise: Promise<FetchEventData>) {
    const startTime = Date.now();

    const {
      query,
      variables,
      error,
      selections,
      executionResult,
      cacheSnapshot,
      type,
    } = await dataPromise;

    const fetchTime = Date.now() - startTime;

    console.groupCollapsed(
      ...format(
        ['GraphQL ', 'color: gray'],
        [type + ' ', `color: ${error ? 'red' : '#03A9F4'}; font-weight: bold`],
        [`(${fetchTime}ms)`, 'color: gray'],
        [` ${selections.length} selections`, 'color: gray'],

        error && [
          'FAILED',
          'margin-left: 10px; border-radius: 2px; padding: 2px 6px; background: #e84343; color: white',
        ]
      )
    );

    const headerStyles = `font-weight: bold; color: #f316c1`;

    console.group(
      ...format(
        ['Query ', headerStyles],
        ['  ', `background-image: url(https://graphql.org/img/logo.svg)`]
      )
    );

    if (variables) {
      console.log(
        ...format(['Variables', 'color: #25e1e1']),
        stringifyJSONIfEnabled(variables)
      );
    }

    console.log(...format([parseGraphQL(query)]));

    console.groupEnd();

    if (error) {
      console.error(...format(['Error', headerStyles]), cleanObject(error));
    } else {
      console.log(
        ...format(['Result', headerStyles]),
        stringifyJSONIfEnabled(executionResult)
      );
    }

    if (options.showSelections) {
      console.groupCollapsed(...format(['Selections', headerStyles]));
      selections.forEach(
        ({ noIndexSelections, selectionsList, type, ...selection }) => {
          console.log(stringifyJSONIfEnabled(cleanObject(selection)));
        }
      );
      console.groupEnd();
    }

    if (options.showCache) {
      console.log(
        ...format(['Cache snapshot', headerStyles]),
        stringifyJSONIfEnabled(cacheSnapshot)
      );
      console.groupEnd();
    }
  }

  function start() {
    const unsubscribe = eventHandler.onFetchSubscribe(onFetch);

    return unsubscribe;
  }

  return {
    start,
    options,
  };
}
