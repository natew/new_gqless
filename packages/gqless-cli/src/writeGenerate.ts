import fs from 'fs';
import { GraphQLSchema } from 'graphql';
import mkdirp from 'mkdirp';
import { dirname, resolve } from 'path';

import { generate, GenerateOptions } from './generate';

async function writeClientCode({
  destinationPath,
  clientCode,
}: {
  clientCode: string;
  destinationPath: string;
}): Promise<void> {
  if (fs.existsSync(destinationPath)) return;

  await fs.promises.writeFile(destinationPath, clientCode, {
    encoding: 'utf-8',
  });
}

async function writeSchemaCode({
  schemaCode,
  destinationPath,
}: {
  schemaCode: string;
  destinationPath: string;
}): Promise<void> {
  const schemaPath = resolve(dirname(destinationPath), './schema.generated.ts');

  if (fs.existsSync(schemaPath)) {
    const existingCode = await fs.promises.readFile(schemaPath, {
      encoding: 'utf-8',
    });

    if (existingCode === schemaCode) return;
  }

  await fs.promises.writeFile(schemaPath, schemaCode, {
    encoding: 'utf-8',
  });
}

export async function writeGenerate(
  schema: GraphQLSchema,
  destinationPath: string,
  generateOptions?: GenerateOptions
) {
  if (!destinationPath.endsWith('.ts')) {
    const err = Error(
      'You have to specify the ".ts" extension, instead, it received: "' +
        destinationPath +
        '"'
    );

    Error.captureStackTrace(err, writeGenerate);

    throw err;
  }

  destinationPath = resolve(destinationPath);

  const [{ clientCode, schemaCode }] = await Promise.all([
    generate(schema, generateOptions),
    mkdirp(dirname(destinationPath)),
  ]);

  await Promise.all([
    writeClientCode({ clientCode, destinationPath }),
    writeSchemaCode({ schemaCode, destinationPath }),
  ]);

  return destinationPath;
}
