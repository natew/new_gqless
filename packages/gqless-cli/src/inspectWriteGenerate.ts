import { existsSync } from 'fs';
import { resolve } from 'path';

import { GenerateOptions } from './generate';
import { getRemoteSchema } from './introspection';
import { writeGenerate } from './writeGenerate';

export async function inspectWriteGenerate({
  endpoint,
  overwrite,
  destination = './src/generated/index.ts',
  generateOptions,
  cli,
  headers,
}: {
  /**
   * GraphQL API endpoint
   *
   * @example 'http://localhost:3000/graphql'
   */
  endpoint: string;
  /**
   * Whether the generation should overwrite if file already exists
   */
  overwrite?: boolean;
  /**
   * File path destination
   * @default './src/generated/index.ts'
   */
  destination?: string;
  /**
   * Specify generate options
   */
  generateOptions?: GenerateOptions;
  /**
   * Whether it's being called through the CLI
   */
  cli?: boolean;
  /**
   * Specify headers for the introspection HTTP request
   */
  headers?: Record<string, string>;
}) {
  destination = resolve(destination);

  const schema = await getRemoteSchema(endpoint, {
    headers,
  });

  if (!overwrite && existsSync(destination)) {
    const err = Error(
      `File '${destination}' already exists, specify ${
        cli ? "'--overwrite'" : "'overwrite: true'"
      } to overwrite the existing file.`
    );

    Error.captureStackTrace(err, inspectWriteGenerate);

    throw err;
  }

  const generatedPath = await writeGenerate(
    schema,
    destination,
    generateOptions
  );

  if (cli) {
    console.log('Code generated successfully at ' + generatedPath);
  }
}
