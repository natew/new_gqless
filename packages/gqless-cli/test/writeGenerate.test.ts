import fs from 'fs';
import path from 'path';
import { createTestApp } from 'test-utils';

import { writeGenerate } from '../src/writeGenerate';
import { getTempDir } from './utils';

const { server, isReady } = createTestApp({
  schema: `
    type Query {
        hello: String!
    }
    `,
  resolvers: {
    Query: {
      hello() {
        return 'hello world';
      },
    },
  },
});

beforeAll(async () => {
  await isReady;
});

test('generates code and writes existing file', async () => {
  const tempDir = await getTempDir({
    initSchemaFile: "console.log('hello world')",
  });

  try {
    const shouldBeIncluded = '// This should be included';

    const firstStats = await fs.promises.stat(tempDir.schemaPath);

    await writeGenerate(server.graphql.schema, tempDir.clientPath, {
      preImport: shouldBeIncluded,
    });

    const secondStats = await fs.promises.stat(tempDir.schemaPath);

    expect(secondStats.mtimeMs).toBeGreaterThan(firstStats.mtimeMs);

    // If the code didn't change, it shouldn't write anything
    await writeGenerate(server.graphql.schema, tempDir.clientPath, {
      preImport: shouldBeIncluded,
    });

    const thirdStats = await fs.promises.stat(tempDir.schemaPath);

    expect(secondStats.mtimeMs).toBe(thirdStats.mtimeMs);

    const generatedContent = await fs.promises.readFile(tempDir.schemaPath, {
      encoding: 'utf-8',
    });

    expect(
      generatedContent
        .split('\n')
        .slice(3)
        .join('\n')
        .startsWith(shouldBeIncluded)
    ).toBeTruthy();

    expect(generatedContent).toMatchSnapshot('overwrite');
  } finally {
    await tempDir.cleanup();
  }
});

test('creates dir, generates code and writes new file', async () => {
  const tempDir = await getTempDir();

  try {
    const targetPath = path.join(
      tempDir.clientPath,
      '/new_path/file-to-generate.ts'
    );

    const shouldBeIncluded = '// This should be included';

    const destinationPath = await writeGenerate(
      server.graphql.schema,
      targetPath,
      {
        preImport: shouldBeIncluded,
      }
    );

    const generatedContentSchema = await fs.promises.readFile(
      path.resolve(path.dirname(destinationPath), './schema.generated.ts'),
      {
        encoding: 'utf-8',
      }
    );

    expect(
      generatedContentSchema
        .split('\n')
        .slice(3)
        .join('\n')
        .startsWith(shouldBeIncluded)
    ).toBeTruthy();

    expect(generatedContentSchema).toMatchSnapshot('new schema');

    const generatedContentClient = await fs.promises.readFile(
      path.resolve(path.dirname(destinationPath), './file-to-generate.ts'),
      {
        encoding: 'utf-8',
      }
    );

    expect(generatedContentClient).toMatchSnapshot('new client');
  } finally {
    await tempDir.cleanup();
  }
});

test('generates code and writes existing file', async () => {
  const tempDir = await getTempDir({
    clientFileName: './client.js',
  });

  try {
    try {
      await writeGenerate(server.graphql.schema, tempDir.clientPath);

      throw Error("shouldn't react");
    } catch (err: unknown) {
      expect(err).toEqual(
        Error(
          `You have to specify the ".ts" extension, instead, it received: "${tempDir.clientPath}"`
        )
      );
    }
  } finally {
    await tempDir.cleanup();
  }
});
