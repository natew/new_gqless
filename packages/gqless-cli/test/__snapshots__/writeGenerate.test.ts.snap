// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates dir, generates code and writes new file: new client 1`] = `
"import { createClient, QueryFetcher } from '@dish/gqless';
import {
  GeneratedSchema,
  generatedSchema,
  scalarsEnumsHash,
} from './schema.generated';

const queryFetcher: QueryFetcher = async function (query, variables) {
  const response = await fetch('/graphql', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query,
      variables,
    }),
    mode: 'cors',
  });

  if (!response.ok) {
    throw new Error(\`Network error, received status code \${response.status}\`);
  }

  const json = await response.json();

  return json;
};

export const client = createClient<GeneratedSchema>({
  schema: generatedSchema,
  scalarsEnumsHash,
  queryFetcher,
});

export const {
  query,
  mutation,
  mutate,
  subscription,
  resolved,
  refetch,
} = client;

export * from './schema.generated';
"
`;

exports[`creates dir, generates code and writes new file: new schema 1`] = `
"// This should be included

import { ScalarsEnumsHash, SchemaUnionsKey } from '@dish/gqless';

export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export interface Scalars {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
}

export const scalarsEnumsHash: ScalarsEnumsHash = {
  String: true,
  Boolean: true,
};
export const generatedSchema = {
  query: { __typename: { __type: 'String!' }, hello: { __type: 'String!' } },
  mutation: {},
  subscription: {},
} as const;

export interface Query {
  __typename: 'Query';
  hello: ScalarsEnums['String'];
}

export interface Mutation {
  __typename: 'Mutation';
}

export interface Subscription {
  __typename: 'Subscription';
}

export interface GeneratedSchema {
  query: Query;
  mutation: Mutation;
  subscription: Subscription;
}

export type MakeNullable<T> = {
  [K in keyof T]: T[K] | null;
};

export interface ScalarsEnums extends MakeNullable<Scalars> {}
"
`;

exports[`generates code and writes existing file: overwrite 1`] = `
"// This should be included

import { ScalarsEnumsHash, SchemaUnionsKey } from '@dish/gqless';

export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export interface Scalars {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
}

export const scalarsEnumsHash: ScalarsEnumsHash = {
  String: true,
  Boolean: true,
};
export const generatedSchema = {
  query: { __typename: { __type: 'String!' }, hello: { __type: 'String!' } },
  mutation: {},
  subscription: {},
} as const;

export interface Query {
  __typename: 'Query';
  hello: ScalarsEnums['String'];
}

export interface Mutation {
  __typename: 'Mutation';
}

export interface Subscription {
  __typename: 'Subscription';
}

export interface GeneratedSchema {
  query: Query;
  mutation: Mutation;
  subscription: Subscription;
}

export type MakeNullable<T> = {
  [K in keyof T]: T[K] | null;
};

export interface ScalarsEnums extends MakeNullable<Scalars> {}
"
`;
